nova/virt/vmwareapi/volumeops.py

```diff
@@ -608,3 +609,120 @@ def attach_root_volume(self, connection_info, instance,
+        field, which is a managed-object reference to a shadow-vm
+        """
+        # This should sensibly moved out to cinder:
+        # Cinder can delete the old shadow-vm, as soon as the attachment
+        # for the vm prior the vmotion gets deleted
+        vm_ref = vm_util.get_vm_ref(self._session, instance)
+        for device in vm_util.get_hardware_devices(self._session, vm_ref):
+            class_name = device.__class__.__name__
+
+            if class_name != "VirtualDisk":
+                continue
+
+            data = shadow_vms.get(device.key)
+            if not data:
+                continue
+
+            try:
+                current_device_path = device.backing.fileName
+                volume = data["volume"]
+                volume_ref = vutil.get_moref(volume, "VirtualMachine")
+                original_device = self._get_vmdk_base_volume_device(volume_ref)
+                if original_device:
+                    original_device_path = original_device.backing.fileName
+                    if original_device_path == current_device_path:
+                        # Already attached
+                        continue
+
+                    # That should not happen
+                    LOG.warning("Shadow-vm %s already has a disk"
+                        " attached at %s replacing it with %s",
+                        volume, original_device_path, current_device_path,
+                        instance=instance
+                        )
+                    self.detach_disk_from_vm(self, volume_ref, instance,
+                        original_device, destroy_disk=True)
+
+                disk_type = vm_util._get_device_disk_type(device)
+                self.attach_disk_to_vm(volume_ref,
+                    instance,
+                    constants.DEFAULT_ADAPTER_TYPE,
+                    disk_type,
+                    current_device_path
+                    )
+            except Exception:
+                LOG.exception("Failed to attach volume {}. Device {}".format(
+                    data["volume_id"],
+                    device.key), instance=instance)
+
+    def delete_shadow_vms(self, block_device_info, instance=None):
+        # We need to delete the migrated shadow vms
+        # (until we implement it in cinder)
+        block_device_mapping = driver.block_device_info_get_mapping(
+            block_device_info)
+
+        if not block_device_mapping:
+            return
+
+        session = self._session
+        deleted = []
+        for disk in block_device_mapping:
+            connection_info = disk["connection_info"]
+            try:
+                data = connection_info["data"]
+                volume_ref = self._get_volume_ref(data["volume"])
+                destroy_task = session._call_method(session.vim,
+                                            "Destroy_Task",
+                                            volume_ref)
```

Comment:
> Does this work with a VM missing its vmdk? 

Not sure, haven't run into that case yet.
>  Destroy a image-cache template, because (iirc) it was missing its vmdk. Had to delete the datastore folder and unregister it instead.

We might have to consider then to make a function for that in general.

> My assumption is, that the disk the shadow-vm points to is no longer available because we migrated it together with the instance.

It works a bit differently: We migrated the attached volume of the "proper" instance will all backings to the new vcenter. But the shadow-vm also keeps a reference to the same file, which causes the source file to stay.
Which is, in a way, great, because we have the backup in case anything goes wrong.
Iit will get deleted with the shadow-vm in question after the migration has completed successfully.

> Is there any reason we cannot use vm_util.destroy_vm() with a given vm_ref parameter here?

Yes, it would print misleading log messages about the vm being destroyed, while the shadow-vm is.